src/lib/coerce.js

```diff
@@ -361,58 +361,81 @@ exports.valObjectMeta = {
  *      as a convenience, returns the value it finally set
  */
 exports.coerce = function(containerIn, containerOut, attributes, attribute, dflt) {
-    var opts = nestedProperty(attributes, attribute).get();
+    return _coerce(containerIn, containerOut, attributes, attribute, dflt).val;
+};
+
+function _coerce(containerIn, containerOut, attributes, attribute, dflt, opts) {
+    var shouldValidate = (opts || {}).shouldValidate;
+
+    var attr = nestedProperty(attributes, attribute).get();
+    if(dflt === undefined) dflt = attr.dflt;
+    var src = ''; // i.e. default
+
     var propIn = nestedProperty(containerIn, attribute);
     var propOut = nestedProperty(containerOut, attribute);
-    var v = propIn.get();
+    var valIn = propIn.get();
 
     var template = containerOut._template;
-    if(v === undefined && template) {
-        v = nestedProperty(template, attribute).get();
+    if(valIn === undefined && template) {
+        valIn = nestedProperty(template, attribute).get();
+        if(valIn !== undefined && shouldValidate && validate(valIn, attr)) src = 't'; // template
```

Comment:
That said, is this particular line actually necessary? Since this step is overwriting `valIn`, after this we're going to treat it as though it came from the container, so on line 406 (or 398) we'll reset it to `'c'`. In practice we don't care about this for now - given the `coerce2` logic all we care about is truthy (specified) or falsy (default), but in case that ever changes, seems like we should either:
- remove this line and pick a single truthy value for `src` so future users don't think `'c'` vs `'t'` is accurate
- or fix lines 398 and 406 to preserve a pre-existing `'t'`, something like:
```js
// 398
src: src || 'c'
//406
if(!src && valOut !== undefined && shouldValidate && validate(valIn, attr)) src = 'c';
```

Reply:
Good call. Revised in https://github.com/plotly/plotly.js/pull/4904/commits/24d1da9d3d134a28523622bfa4275fc43f9aed37.